fuel-merkle/src/tests/binary_verify.rs

```diff
@@ -0,0 +1,164 @@
+    fn random_tree(min: usize, max: usize)(values in values(min, max)) -> (Vec<Value>, MerkleTree<TestTable, StorageMap<TestTable>>) {
+        let storage = StorageMap::<TestTable>::new();
+        let mut tree = MerkleTree::new(storage);
+        for datum in values.iter() {
+            tree.push(datum.as_ref()).unwrap();
+        }
+        (values, tree)
+    }
+}
+
+proptest! {
+    #[test]
+    fn verify__returns_true_for_valid_proof((values, tree) in random_tree(1, 1_000), arb_num: usize){
+        let num_leaves = values.len();
+        let index = arb_num % num_leaves;
+        let data = values[index];
+
+        // Given
+        let (root, proof_set) = tree.prove(index  as u64).expect("Unable to generate proof");
+
+        // When
+        let verification = verify(&root, &data, &proof_set, index as u64, num_leaves  as u64);
+
+        // Then
+        prop_assert!(verification)
+    }
+
+    #[test]
+    fn verify__returns_false_for_invalid_root((values, tree) in random_tree(1, 1_000), arb_num: usize, root: Bytes32){
+        let num_leaves = values.len();
+        let index = arb_num % num_leaves;
+        let data = values[index];
+
+        // Given
+        let (_, proof_set) = tree.prove(index  as u64).expect("Unable to generate proof");
+
+        // When
+        let verification = verify(&root, &data, &proof_set, index as u64, num_leaves  as u64);
+
+        // Then
+        prop_assert!(!verification)
+    }
+
+    #[test]
+    fn verify__returns_false_for_invalid_proof_set((values, tree) in random_tree(1, 1_000), arb_num: usize, proof_set: ProofSet){
+        let num_leaves = values.len();
+        let index = arb_num % num_leaves;
+        let data = values[index];
+
+        // Given
+        let (root, _) = tree.prove(index  as u64).expect("Unable to generate proof");
+
+        // When
+        let verification = verify(&root, &data, &proof_set, index as u64, num_leaves  as u64);
+
+        // Then
+        prop_assert!(!verification)
+    }
+
+    #[test]
+    fn verify__returns_true_for_valid_proof_of_last_leaf((values, tree) in random_tree(1, 1_000)){
+        let num_leaves = values.len();
+        let index = num_leaves - 1;
+        let data = values[index];
+
+        // Given
+        let (root, proof_set) = tree.prove(index as u64).expect("Unable to generate proof");
+
+        // When
+        let verification = verify(&root, &data, &proof_set, index as u64, num_leaves as u64);
+
+        // Then
+        prop_assert!(verification)
+    }
+
+    #[test]
+    fn verify__returns_false_for_invalid_proof_of_last_leaf((values, tree) in random_tree(1, 1_000)){
+        let num_leaves = values.len();
+        let index = num_leaves - 1;
+        let data = values[index];
+
+        // Given
+        let (root, proof_set) = tree.prove(index as u64).expect("Unable to generate proof");
+
+        // When
+        let verification = verify(&root, &data, &proof_set, index as u64, num_leaves as u64 + 1);
```

Comment:
It would be nice to have random `num_leaves` here as well

Reply:
Hmm, I made this change locally but proptest is reporting failures. I will need to dig into this.

Reply:
What failures do you have? Is that in cases when `num_leaves` is more than the real number of leaves?